lib/object.c

```diff
@@ -42,14 +210,99 @@ static tool_rc tpm2_util_object_load2(ESYS_CONTEXT *ctx, const char *objectstr,
+            tpk = PEM_read_bio_TSSPRIVKEY_LOAD(input_bio, NULL, NULL, NULL);
+            if (tpk == NULL) {
+                LOG_ERR("Unable to read PEM from provided BIO/file");
+                goto error;
+            }
+            int pub_len = tpk->pubkey->length;
+            int priv_len = tpk->privkey->length;
+
+            if (Tss2_MU_TPM2B_PUBLIC_Unmarshal(tpk->pubkey->data, pub_len,
+                                                NULL, pub)) {
+                                                    LOG_ERR("Error deserializing public portion of object");
+                                                    goto error;
+                                                }
+            if (Tss2_MU_TPM2B_PRIVATE_Unmarshal(tpk->privkey->data, priv_len,
+                                                NULL, priv)) {
+                                                    LOG_ERR("Error deserializing private portion of object");
+                                                    goto error;
+                                                }
+            uint64_t val;
+            int ret = ASN1_INTEGER_get_uint64(&val, tpk->parent);
+            if (ret == 0 || val > UINT32_MAX) {
+                LOG_ERR("Unable to convert parent to integer/Value too large");
+                goto error;
+            }
+            
+            //if parent == TPM2_RH_OWNER or 0 then non-persistent parent
+            bool is_persistent_parent = (val != TPM2_RH_OWNER && val != 0);
+            if (!is_persistent_parent) {
+                ESYS_TR obj_parent = ESYS_TR_NONE;
+                rc = setup_primary(ctx, &obj_parent);
+                if (rc != tool_rc_success) {
+                    LOG_ERR("Unable to create parent using create primary");
+                    goto error;
+                }
+                outobject->tr_handle = obj_parent;
+
+                TSS2_RC ret = Esys_TR_GetTpmHandle(ctx, obj_parent, &outobject->handle);
+                if (ret != TSS2_RC_SUCCESS) {
+                    LOG_ERR("Unable to fetch TPM handle from ESYS TR handle");
+                    goto error;
+                }
+            } else {
+                ESYS_TR tr_parent = ESYS_TR_NONE;
+                rc = tpm2_from_tpm_public(ctx, val, ESYS_TR_NONE, ESYS_TR_NONE,
+                                        ESYS_TR_NONE, &tr_parent);
+                if (rc != tool_rc_success) {
+                    LOG_ERR("Unable to fetch TR Handle for persistent parent");
+                    goto error;
+                }
+                outobject->tr_handle = tr_parent;
+            }
+            /*tpm2_session *sess_handle = NULL;
+            rc = tpm2_auth_util_from_optarg(ctx, 0, &sess_handle, false);
+            if (rc != tool_rc_success) {
+                LOG_ERR("Unable to get session handle for parent");
+                goto error;
+            }
+            outobject->session = sess_handle;*/
```

Comment:
@idesai @williamcroberts This is one of the spots in object.c (which I had added) where there was a leak. I am already fetching the session above (line 196) and then fetching it again here. If I understand correctly, the function call "tpm2_auth_util_from_optarg" allocates memory for the session handle each time its called.